include/albatross/src/core/linear_combination.hpp

```diff
@@ -74,6 +74,63 @@ inline auto difference(const X &x, const Y &y) {
   return difference(vx, vy);
 }
 
+template <typename Derived, typename X>
+inline auto operator*(const Eigen::SparseMatrixBase<Derived> &matrix,
+                      const std::vector<X> &features) {
+  std::vector<LinearCombination<X>> output(cast::to_size(matrix.rows()));
+
+  std::vector<std::vector<double>> output_coefs(cast::to_size(matrix.rows()));
+
+  for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
+    for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
+      output[cast::to_size(it.row())].values.push_back(
+          features[cast::to_size(it.col())]);
+      output_coefs[cast::to_size(it.row())].push_back(it.value());
+    }
+  }
+
+  for (std::size_t i = 0; i < output_coefs.size(); ++i) {
+    output[i].coefficients = Eigen::Map<Eigen::VectorXd>(
+        &output_coefs[i][0], cast::to_index(output_coefs[i].size()));
+  }
+
+  return output;
+}
+
+template <typename Derived, typename X>
+inline auto operator*(const Eigen::SparseMatrixBase<Derived> &matrix,
+                      const std::vector<LinearCombination<X>> &features) {
+  std::vector<LinearCombination<X>> output(cast::to_size(matrix.rows()));
+
+  std::vector<std::vector<double>> output_coefs(cast::to_size(matrix.rows()));
+
+  for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
+    for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
+      const auto &combo = features[cast::to_size(it.col())];
+      for (Eigen::Index i = 0; i < combo.coefficients.size(); ++i) {
+        output[cast::to_size(it.row())].values.push_back(
+            combo.values[cast::to_size(i)]);
+        output_coefs[cast::to_size(it.row())].push_back(it.value() *
```

Comment:
If we can output directly to the coefficient vector as suggested above in this function as well, this loop can go away entirely and we can just multiply `combo.coefficients` by `it.value()`.

Reply:
Mind putting a code snippet suggestion up? I'm not sure I follow. Effectively what we're doing here is concatenating all the linear combo values and coeffs together:
```
  output[i].values = concatenate(f[i_0].values, ..., f[i_n].vlaues);
  output[i].coefficients = concatenate(w[i_0] * coefs[i_0], ... , w[i_n] * coefs[i_n]);
  // i_j indicates the j^th non zero index in row i
```
So I can see how we could first loop over and learn the size of the output coefs, allocate that and copy (with a multiplication) into the `Eigen::VectorXd` with blocks ... but I don't see how loops go away unless something really creative happens inside the `concatenate` place holders? Maybe that's what you meant by just multiplying coefs by the value?

Reply:
This passes all tests for me:

```
@@ -102,26 +102,16 @@ inline auto operator*(const Eigen::SparseMatrixBase<Derived> &matrix,
                       const std::vector<LinearCombination<X>> &features) {
   std::vector<LinearCombination<X>> output(cast::to_size(matrix.rows()));
 
-  std::vector<std::vector<double>> output_coefs(cast::to_size(matrix.rows()));
-
   for (Eigen::Index k = 0; k < matrix.outerSize(); ++k) {
     for (typename Derived::InnerIterator it(matrix.derived(), k); it; ++it) {
       const auto row = cast::to_size(it.row());
       const auto col = cast::to_size(it.col());
       const auto &combo = features[col];
-      for (Eigen::Index i = 0; i < combo.coefficients.size(); ++i) {
-        const auto si = cast::to_size(i);
-        output[row].values.push_back(combo.values[si]);
-        output_coefs[row].push_back(it.value() * combo.coefficients[i]);
-      }
+      output[row].values = combo.values;
+      output[row].coefficients = it.value() * combo.coefficients;
     }
   }
 
-  for (std::size_t i = 0; i < output_coefs.size(); ++i) {
-    output[i].coefficients = Eigen::Map<Eigen::VectorXd>(
-        &output_coefs[i][0], static_cast<Eigen::Index>(output_coefs[i].size()));
-  }
-
   return output;
 }
```

Reply:
I don't think that's doing the same thing, if I'm correct that ^ will fill `output[row]` with only the product from the last entry in that `row`. I guess this means there's a gap in the tests (they mostly just check the output types, probably should add some validation on the values).

Reply:
I've updated the tests to actually check the content of the transformations after which that ^ no longer passes, feel free to experiment if you think there's a better way (but if you find yourself looking at this chart https://xkcd.com/1691/ ...).

Reply:
It took me a while of staring at it, but I get it now.  Sorry for the confusion.  Glad the tests are better at least!